PMax Assets: Support editing the Business name asset

eason9487 opened the pull request:
### Changes proposed in this Pull Request:

This PR relates to https://github.com/woocommerce/google-listings-and-ads/issues/1787 and implements the missing Business name asset.

- Let `AppButton` support the `hidden` attribute.
- Make related modules support the single text value:
   - Update `validateAssetGroup` to check it.
   - `AssetGroupCard` handle it with `TextsEditor`.
   - Hide the Add button in `TextsEditor` when the min and max limits are the same and bigger than 1.
- Support editing the business name asset in the `AssetGroupCard`.

### Screenshots:

![image](https://user-images.githubusercontent.com/17420811/221552698-37215beb-a5c0-4451-b23e-608d1f187e8a.png)

![image](https://user-images.githubusercontent.com/17420811/221552776-bfdb82c7-9d6d-4e5d-a88a-ec883ff5acd4.png)

![image](https://user-images.githubusercontent.com/17420811/221553025-5a93d635-6a37-4c82-b569-f255a32fc5f8.png)

![image](https://user-images.githubusercontent.com/17420811/221552814-dc64d673-383b-4328-bd48-3209bdba5a43.png)

![image](https://user-images.githubusercontent.com/17420811/221552861-1b396bf0-126e-4b80-9bb8-031b3c6ab22b.png)

![image](https://user-images.githubusercontent.com/17420811/221552895-9090c995-4853-48d1-b720-0bad46185ffd.png)

### Detailed test instructions:

#### 💡 Additional notes before testing

Google Ads side will check if the domain in the Final URL matches the claimed domain of the linked Merchant Center account. Therefore, it may need to make the local env have the same domain by switching the local env to be accessible publicly or replacing the domain temporarily. The way I used:
- Replace [this line](https://github.com/woocommerce/google-listings-and-ads/blob/2f448130ce83497e46102a63e28464505f30e146/js/src/components/paid-ads/convertToAssetGroupUpdateBody.js#L86) with:
   ```js
   final_url: values.final_url.replace( 'your local domain', 'GMC domain' ),
   ```
- Add the following at [this line](https://github.com/woocommerce/google-listings-and-ads/blob/2f448130ce83497e46102a63e28464505f30e146/src/Ads/AssetSuggestionsService.php#L191):
   ```php
   $final_url = str_replace( 'your local domain', 'GMC domain', $final_url );
   ```
- Add the following at the end of **google-listings-and-ads.php** if got the "cURL error 60" issue with `wp_remote_get`:
   ```php
   add_filter( 'https_ssl_verify', '__return_false' );
   add_filter( 'https_local_ssl_verify', '__return_false' );
   ```

#### Testing

1. Go to step 2 of the campaign creation or editing page.
2. Check if the Business name field works well.
3. Check if the Business name is able to update.

### Additional details:

The discussion of this change: pcTzPl-12o-p2#comment-2294

### Changelog entry

eason9487 committed:
Let AppButton support the `hidden` attribute

eason9487 committed:
Let `validateAssetGroup` support the single text value.

eason9487 committed:
Hide the Add button in `TextsEditor` when the when the min and max limits are the same and bigger than 1.

eason9487 committed:
Let `AssetGroupCard` handle the single text value with `TextsEditor`.

eason9487 committed:
Support editing the business name asset in the `AssetGroupCard`.

puntope commented on js/src/components/paid-ads/asset-group/asset-group-card.js:
```diff
@@ -126,8 +126,8 @@ export default function AssetGroupCard() {
-			{ ASSET_TEXT_SPECS.map( ( spec, index ) => {
-				const initialTexts = baseAssetGroup[ spec.key ];
+			{ ASSET_TEXT_SPECS.map( ( spec ) => {
+				const initialTexts = [ baseAssetGroup[ spec.key ] ].flat();
```
As far as I know this would be also impacted by #1907

eason9487 replied:
A code conflict has been resolved in https://github.com/woocommerce/google-listings-and-ads/pull/1908/commits/1759264a5e55e5763850dca8c051bb7c516b9a2b.

puntope commented on js/src/components/paid-ads/validateAssetGroup.js:
```diff
@@ -75,12 +75,13 @@ export default function validateAssetGroup( values ) {
-				spec.min,
-				name
-			);
+			const format = spec.requiredSingleValue
+				? // translators: 1: (This variable is omitted). 2: Asset field name.
+				  __( 'The %2$s is required', 'google-listings-and-ads' )
+				: // translators: 1: The minimal number of this item. 2: Asset field name.
+				  __( 'Add at least %1$d %2$s', 'google-listings-and-ads' );
+
+			const message = sprintf( format, spec.min, name );
```
```jsx
const message = spec.requiredSingleValue 
             ? sprintf(
				// translators: 1: Asset field name.
				__( 'The %1$s is required', 'google-listings-and-ads' ),
				name
			)
            :  sprintf(
				// translators: 1: The minimal number of this item. 2: Asset field name.
				__( 'Add at least %1$d %2$s', 'google-listings-and-ads' ),
				spec.min,
				name
			);
```

💅  I'd suggest this version avoiding that `(This variable is omitted)` that could be confusing for translators.

puntope replied:
Another option could be to revert the fields order 

```jsx
const format = spec.requiredSingleValue
				? // translators: 1 : Asset field name.
				  __( 'The %1$s is required', 'google-listings-and-ads' )
				: // translators: 1:  Asset field name. 2: The minimal number of this item
				  __( 'Add at least %2$d %1$s', 'google-listings-and-ads' );

			const message = sprintf( format, name, spec.min );
```

eason9487 replied:
Thanks for the great suggestion. Adjusted in https://github.com/woocommerce/google-listings-and-ads/pull/1908/commits/32f9f8b295ce33ab0b943e1e0c51c893c2ed1253.

puntope commented on js/src/components/paid-ads/asset-group/asset-group-card.js:
```diff
@@ -138,14 +138,15 @@ export default function AssetGroupCard() {
-										) }
-									</ExternalLink>
-								) }
+								{ spec.key === ASSET_FORM_KEY.HEADLINE &&
+									isSelectedFinalUrl && (
+										<ExternalLink href="https://support.google.com/google-ads/answer/6167101">
+											{ __(
+												'Learn how to write effective ads',
+												'google-listings-and-ads'
+											) }
+										</ExternalLink>
+									) }
 							</>
```
💅  We could potentially add this in the specs as props. (I mean the link and the link label) 

```jsx
 { spec.linkUrl && spec.linkLabel && isSelectedFinalUrl && (
        <ExternalLink href={linkUrl}>
...
```

eason9487 replied:
Good idea. I moved all JSX of the link to the `assetSpecs` module by https://github.com/woocommerce/google-listings-and-ads/pull/1908/commits/465f11ea01c8d09dba5618ba12a404c58227ea54.

puntope reviewed (commented):
Howdy @eason9487 

Overall testing good. The only blocker is the issue in this comment

https://github.com/woocommerce/google-listings-and-ads/pull/1908#discussion_r1119000221

eason9487 committed:
Reverse the parameters call to `sprintf` in validateAssetGroup to avoid confusion for translators.

Address https://github.com/woocommerce/google-listings-and-ads/pull/1908#discussion_r1119015536

eason9487 committed:
Move the additional link in the subheading of the Headline field to the assetSpecs module.

Address https://github.com/woocommerce/google-listings-and-ads/pull/1908#discussion_r1119025792

eason9487 committed:
Merge branch 'feature/pmax-assets' into update/1787-add-business-name

eason9487 commented:
@puntope, thanks for the review. I have addressed all review comments. Could you take another look?

puntope reviewed (approved):
Thanks for the adjustments. Looking good.
